### PR TITLE
Add all color palettes to select from editor panel

### LINKED ISSUE #64734: Placeholders to render theme colors alongside custom colors
### Description

Having in mind that, probably, the theme colors are “more important” than the personalized colors, when you have both, blocks only show the personalized colors by default in the block (for example, in the Cover block), although in the sidebar all colors appears correctly.

Tested in WordPress 6.6.

### Step-by-step reproduction instructions

1. Have some theme colors (not personalized colors).
2. Add a Cover block. It shows to pick one theme color.
3. Add some personalized colors.
4. Add a cover block. It only shows the personalized colors in the block. In the sidebar are all the colors.

### Screenshots, screen recording, code snippet

Cover only with personalized colors
Text only with personalized colors
Cover Sidebar with all the colors

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

- [X] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [X] Yes

### LINKED ISSUE #64857: Block Editor Color Palette: Only custom color palettes are displayed
### Description

If the custom color palette is defined in the global styles, the color palette in the block editor package only shows that custom color palette.

_Note that I'm talking about [the `ColorPalette` component in the block editor package](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/color-palette/), not [the `ColorPalette` component in the components package](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/color-palette)._

This issue can be seen in color inline formatting and Cover block placeholder. From what I've tested in Playground, this problem appears to have been occurring since at least WP6.3. As I recall, all palettes - theme, default, and custom - were displayed in the past.

### Step-by-step reproduction instructions

- Add a custom color palette with the global styles.
- Add a cover block or apply inline color to the text.

### Screenshots, screen recording, code snippet

Cover block and text color format. Only custom colors are shown:

custom-color

If no custom color is defined, the theme color is displayed:

image

Like color block support, it should display all three colors `default`, `theme`, and `custom`:

image

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

- [X] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [X] Yes

### PR DESCRIPTION
**Pull request by AKSHAT2802**
Closes: https://github.com/WordPress/gutenberg/issues/64857
Closes: https://github.com/WordPress/gutenberg/issues/64734
## What?
This PR Adds all the available colors in the list of editor color panel ( default, theme, custom )  so that the user can any predefined color in the editor.

## Why?
When selecting a color we should have a list of all the predefined options for a user instead of only custom-defined options.

## How?
Modified with-color-context.js file to list all the color instead of only custom colors.

### Before
Screenshot 2024-09-09 at 5 07 46 PM

### After

https://github.com/user-attachments/assets/895492b8-0601-461d-affb-cd6361c5ae5e

### CODE REVIEW

**Comment by ndiego** [comment]
Thanks for working on this @AKSHAT2802. However, I don't think we should display the default color palette as well, only the theme color palette in addition to any custom options set by the user. Using the Twenty Twenty-Four theme as an example, this is what it should look like. 

|  There are no custom colors | The user has added custom colors |
|-|-|
|image|image|

**Review by richtabor — CHANGES_REQUESTED** [review]
Two notes: 

1. I should only see the palettes which I have enabled. If I don't have the default palette enabled via theme.json, then I should not see that palette here.

2. I'd expect the colors in the same order as the color palette elsewhere, with theme colors first, then custom colors. In the screenshot below I'd expect to see my theme colors, then the custom red color: 
CleanShot 2024-11-07 at 09 19 10

**Comment by AKSHAT2802** [comment]
Hii @richtabor & @ndiego 
Thanks for the review 
Updated the PR

Sample Images 

1] Theme palette with custom colors
image

2] Changed theme palette with no custom colors
image

3] Changed theme palette and added custom colors
image

File: packages/block-editor/src/components/color-palette/with-color-context.js [hunk 1]
```diff
@@ -10,14 +10,17 @@ import { useSettings } from '../use-settings';
...
+			'color.palette.theme',
+			'color.palette.custom',
 			'color.custom'
 		);
-		const {
-			colors = colorsFeature,
-			disableCustomColors = ! enableCustomColors,
-		} = props;
+
+		const colors = [ ...( themeColors || [] ), ...( customColors || [] ) ];
```
Comment (t-hamano) [inline packages/block-editor/src/components/color-palette/with-color-context.js]:
```suggestion
		const colors = [ ...themeColors, ...defaultColors, ...customColors ];
```

This simplification should be fine.

**Review by t-hamano — COMMENTED** [review]
I think the default palette should not be disabled all the time, but only when `settings.color.defaultPalette` is `false`, otherwise it won't match other color palettes. Moreover, the order is "Theme" > "Default" > "Custom":

## Current

image

## Expected

image

Additionally, don't forget to test the inline text highlighting palette in this PR:

image

**Comment by richtabor** [comment]
> I think the default palette should not be disabled all the time, but only when settings.color.defaultPalette is false, otherwise it won't match other color palettes. Moreover, the order is "Theme" > "Default" > "Custom":

+1

**Comment by AKSHAT2802** [comment]
Thanks, @t-hamano for the update. 
I have changed the code according to the requirements.

File: packages/block-editor/src/components/color-palette/with-color-context.js [hunk 2]
```diff
@@ -10,14 +10,27 @@ import { useSettings } from '../use-settings';
...
-		const {
-			colors = colorsFeature,
-			disableCustomColors = ! enableCustomColors,
-		} = props;
+
+		const colors = enableDefaultColors
+			? themeColors.concat( defaultColors, customColors )
+			: themeColors.concat( customColors );
+
+		const { disableCustomColors = ! enableCustomColors } = props;
```
Comment (t-hamano) [inline packages/block-editor/src/components/color-palette/with-color-context.js]:
```suggestion
		const _colors = enableDefaultColors
			? [
					...( themeColors || [] ),
					...( defaultColors || [] ),
					...( customColors || [] ),
			  ]
			: [ ...( themeColors || [] ), ...( customColors || [] ) ];

		const { colors = _colors, disableCustomColors = ! enableCustomColors } =
			props;
```

This change contains two fixes:

- Fixes a critical error that occurs when theme colors or custom colors are missing.
- Respects the `colors` prop of the `ColorPalette` component.

**Comment by t-hamano** [comment]
Thanks for the update.

To ship this PR, we need to resolve the unit test failure. The failing tests are for the Cover block, and I think it's because the PR changed the behavior of the placeholder.

In trunk, the placeholder shows the default palette even if there are no colors available:

image

In this PR, no palette is shown:

image

The behavior in this PR should be correct, so we need to fix the unit tests.

The following changes should solve the problem, could you try them?

```diff
diff --git a/packages/block-library/src/cover/test/edit.js b/packages/block-library/src/cover/test/edit.js
index 5c1a5b5e13..f5d6a5301e 100644
--- a/packages/block-library/src/cover/test/edit.js
+++ b/packages/block-library/src/cover/test/edit.js
@@ -337,7 +337,7 @@ describe( 'Cover block', () => {
                        describe( 'when colors are disabled', () => {
-                                       await createAndSelectBlock();
+                                       await selectBlock( 'Block: Cover' );
                                        await userEvent.click(
                                                screen.getByRole( 'tab', { name: 'Styles' } )
                                        );
@@ -350,7 +350,7 @@ describe( 'Cover block', () => {
                                } );
                                test( 'does not render opacity control', async () => {
                                        await setup( undefined, true, disabledColorSettings );
-                                       await createAndSelectBlock();
+                                       await selectBlock( 'Block: Cover' );
                                        await userEvent.click(
                                                screen.getByRole( 'tab', { name: 'Styles' } )
                                        );
```

**Review by richtabor — APPROVED** [review]
LGTM 
CleanShot 2024-11-21 at 17 48 04

**Comment by t-hamano** [comment]
It has passed all CI checks so let's merge this PR.

@AKSHAT2802 Thank you for your efforts!

**Pull request merged**